parse/src/main/java/com/parse/ParseObjectStoreMigrator.java

```diff
@@ -0,0 +1,82 @@
+package com.parse;
+
+import com.parse.boltsinternal.Continuation;
+import com.parse.boltsinternal.Task;
+import java.util.Arrays;
+
+/** Use this utility class to migrate from one {@link ParseObjectStore} to another */
+class ParseObjectStoreMigrator<T extends ParseObject> implements ParseObjectStore<T> {
+
+    private final ParseObjectStore<T> store;
+    private final ParseObjectStore<T> legacy;
+
+    /**
+     * @param store the new {@link ParseObjectStore} to migrate to
+     * @param legacy the old {@link ParseObjectStore} to migrate from
+     */
+    public ParseObjectStoreMigrator(ParseObjectStore<T> store, ParseObjectStore<T> legacy) {
+        this.store = store;
+        this.legacy = legacy;
+    }
+
+    @Override
+    public Task<T> getAsync() {
+        return store.getAsync()
+                .continueWithTask(
+                        new Continuation<T, Task<T>>() {
+                            @Override
+                            public Task<T> then(Task<T> task) throws Exception {
+                                if (task.getResult() != null) return task;
+                                return legacy.getAsync()
+                                        .continueWithTask(
+                                                new Continuation<T, Task<T>>() {
+                                                    @Override
+                                                    public Task<T> then(Task<T> task)
+                                                            throws Exception {
+                                                        T object = task.getResult();
+                                                        if (object == null) return task;
+                                                        return legacy.deleteAsync()
+                                                                .continueWith(
+                                                                        task1 ->
+                                                                                ParseTaskUtils.wait(
+                                                                                        store
+                                                                                                .setAsync(
+                                                                                                        object)))
+                                                                .onSuccess(task1 -> object);
+                                                    }
+                                                });
+                            }
+                        });
```

Comment:
Is there a better way to structure this code?

Reply:
Will look into it.